### ISSUE TITLE
Include new Adminer version in LAMP/LAPP (and derivative) appliances

### ISSUE THREAD

**Issue by JedMeister** [turnkeylinux/tracker#734]
Apologies on the wordiness of this but I wanted to provide some background and context.

Adminer's MySQL functionality is quite good. However, as some users have noted, it's support of PostgreSQL is pretty minimal. The version that we have been including ([3.3.3](https://packages.debian.org/jessie/adminer)) is very dated and apparently newer versions provide vastly improved PostgreSQL support/features.

I noticed a while ago that the Debian Adminer package was also at risk of being removed from Stretch. There were some build bugs and the previous maintainer had disappeared. I was a bit concerned about what we would do if that happened and considered looking into helping out with the packaging myself. I started playing with that, but the learning curve was steeper than I expected.

I voiced my concerns in general discussions with ace Debian Developer and all round nice guy Chris Lamb (@lamby). Then without explicit request (or payment), he took it upon himself to take over the [maintenance of Adminer in the Debian repos](https://packages.qa.debian.org/a/adminer.html) and upload the current version (4.2.5). That is now in [Stretch](https://packages.debian.org/stretch/adminer) (testing).

Furthermore, he has also kindly backported the current version to [Jessie backports](https://packages.debian.org/jessie-backports/adminer) so we don't need to wait for v15.0 before we can integrate this shiny new version into TurnKey!

Yay, thanks Chris! You're a :star2: ! :smile: 

----

But now we need to give it a bit of a test drive and integrate it into the v14.2 build code. If you have time @DocCyblade perhaps this is a good job for you? Especially seeing as you did the initial Adminer integration! :smile: 

It will require some initial testing before we go any further. I suggest to start with, just manually install it into a v14.1 TKL appliance (actually probably 2; a LAMP based one and a LAPP based one). In case you're no familiar with backports, you'll need to add a sources.list entry. FWIW I personally prefer a separate `.list` file in `/etc/apt/sources.list.d/` as I believe it is more transparent and tidier. At the testing stage though, an additional line in the existing file is fine. After that it should install with:
````
apt-get update && apt-get install -t jessie-backports adminer
````
Assuming that all goes smoothly, then it'll need to be integrated into common. The Adminer conf script will need to add the backports repo and install the package. AFAIK it shouldn't be necessary but as a preventative precaution we should also pin it. We have a few pins already in appliances that you can use a guide. OTTOMH I think asp-net-apace is one?!

**Comment by lamby** [comment]
(No problem!) Let me know if the Debian integratio could be better. in places.

**Comment by DocCyblade** [comment]
@lamby - Thanks for your contribution! At some point in the future I want to dig into and learn debian package creation/maintenance. Do you know if any paths have changed from the version we are using to the new version you packaged?

**Comment by lamby** [comment]
Not that I am aware of.. 

Ken Robinson wrote:

> @lamby - Thanks for your contribution! At some point in the future I want to dig into and learn debian package creation/maintenance. Do you know if any paths have changed from the version we are using to the new version you packaged?
> 
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/turnkeylinux/tracker/issues/734#issuecomment-261607982


-- 
Chris Lamb
chris-lamb.co.uk / @lolamby

**Comment by DocCyblade** [comment]
My plan of attack:

- [ ] Pre-test: download and have go with it on a new install of LAMP/LAPP
   - [ ] LAMP
   - [ ] LAPP
- [ ] Clone and work off new branch from common, for the common bits
   - [ ] Conf scripts
   - [ ] Overlays
- [ ] Test builds using updated common:
   - [ ] LAMP
   - [ ] LAPP

Example Pins:
https://github.com/turnkeylinux-apps/asp-net-apache/blob/master/conf.d/main

**Comment by DocCyblade** [comment]
@JedMeister - Testing is going well, the new version looks good. Will need to make some changes as they added new database types and changed the version checking script. How much do you want me to do? I can update the common bits, and update the lamp/lapp code too so that the back ports gets added. I'll have to see if that can be done in common. It's been too long since I have been mucking about in the common folder

**Comment by JedMeister** [comment]
Hey @DocCyblade - Great work mate! I would love you to do as much as you are happy to do! :smile: 

I really liked the way that you pre-configured the theme and DB type for the different appliances last time. I think that'd be worth doing again if you have the time and energy!? 

FWIW, here's where you put the theme overlay (in common): https://github.com/turnkeylinux/common/tree/master/overlays/adminer/usr/share/adminer/adminer/static

The Apache and Lighty conf files are in there too. I'm pretty sure that the Nginx appliance just has it's Adminer webserver conf in it's own overlay, although IIRC there is at least one other appliance that includes Nginx and Adminer, so we should probably move that to common too. TBH though that's a "if we get there" rather than a "must do".

The conf scripts that adjust the DB (to either MySQL or PostgreSQL) and here:
https://github.com/turnkeylinux/common/blob/master/conf/adminer-mysql
https://github.com/turnkeylinux/common/blob/master/conf/adminer-pgsql

**Comment by DocCyblade** [comment]
@JedMeister thanks for the links! The current theme works with 4.x and we can set local host and server type like we do now!

**Comment by DocCyblade** [comment]
I'll be working on getting pull requests submitted in the coming week

**Comment by DocCyblade** [comment]
@JedMeister - Checking in, found some time to work on this. It looks like however the new version requires the MySQL drive (not sure why) so I can limit the drivers to MySQL for LAMP for sure but PostgreSQL. Looking into maybe overriding one of the Drive variable using and include file to statically set the driver list.

**Comment by DocCyblade** [comment]
@JedMeister - After I just submitted the above comment, I found the below on the plug ins for Adminer should allow us to extend it just the way would like. I am working on implementing these on my test server (https://krajr.tklapp.com:12322)

From https://www.adminer.org/en/plugins/ :

Enable Plug-ins
https://raw.githubusercontent.com/vrana/adminer/master/plugins/plugin.php

Show only localhost
https://raw.githubusercontent.com/vrana/adminer/master/plugins/login-servers.php

Disable version checker
https://raw.githubusercontent.com/vrana/adminer/master/plugins/version-noverify.php

Allows us to do just what we want to do. I may be able to write one to only display the drivers we want.

**Comment by DocCyblade** [comment]
@JedMeister - While mucking around the file system I found (not sure if this is because of the new packager or if it was there all along) the themes are already install by the package as well as the plugins. We may need to just copy them from one directory to another.

**Comment by JedMeister** [comment]
@DocCyblade - nice work mate! :smile: 

That sounds awesome. It may well have been how Chris did the package (better than the previous one - he's an expert Debian Developer!). Or it could just be that because it's a (much) newer version, so all that stuff is now bundled? So it sounds like we'll be able to do a nicer job of the integration for v14.2 AND with a newer, better version!

Regardless, you rock! :smile:

**Comment by DocCyblade** [comment]
@JedMeister - Ok I need your guidance:

After more mucking around, I think we can achieve modifying the login screen and disable the auto checking by [creating a custom plugins](https://www.adminer.org/en/plugins/#create):

After [looking at the API](https://www.adminer.org/en/extension/#api) the login form is all inclusive and I would need to some how hook global variables to do these as separate plugins. I think however I will create LoginForm customizer plugin that will lock in the localhost (server) and lock in database driver. I should be able to even submit it to be included on the [user submitted plugins](https://www.adminer.org/en/plugins/#user). 

Once I have the plugin working, will need to [create a custom loader to enable the plugins](https://www.adminer.org/en/plugins/#use) that loads the plugins that we want, then we have to use that file instead of index.php. 

So how should we go about this. I can create the plug in on a gist and use that in the common config script, or could just add it to the common overlay. (I could always have the original gist location in the file and load the file into common for control purposes) 


As for the index.php we could:

1. Create a index.turnkey.php that includes index.php, then adjust the apache/nginx configs to default to that file instead of index.php.
 
2.  Rename index.php (that could break updates or break our plugins) and use the our new one


I am thinking that 2 could cause issues with updates but would be easy to implement, 1 would be preferred but changes would be needed in all the apache and nginx configs to update the index


Thoughts?




Then [according to the documents](https://www.adminer.org/en/plugins/#use)

**Comment by DocCyblade** [comment]
@JedMeister - Another option would be to just create a plugin for turnkey use so we could also TurnKey brand the login page

**Comment by JedMeister** [comment]
As you point out, all 3 options have pros and cons. And regardless, it actually this sounds like tons more work than I had originally anticipated... That makes you even more of a legend IMO! :smile: 

Your option 1 (or 3) sounds ideal, as it will allow us to configure it how we want, but also ensure it doesn't break with an update. Having said that, I'm not sure if you'd noticed, but for the life of Debian Jessie so far, there has not been a single update! That suggests that it's pretty secure and any issues have been relatively minor. 

In my mind, that means that your option 2 is probably still a legitimate approach. IIRC that's similar to the approach we took last time. I.e. we patched the package managed files. Under normal circumstance that's considered bad practice, but I spoke about it at length with Alon and he was ok with it and to paraphrase Liraz: "so long as you understand the risks of not following convention, there are perfectly legitimate circumstance where breaking the rules is ok". I don't recall the exact context of that remark, so perhaps I'm abusing the spirit of it, but this seems like another scenario where that may apply.

I suggest, that if you are ok to go with option 1/3 then just go for it. If your current workload and family commitments mean that it may be a struggle to get done within the next couple of weeks, then let me know ASAP and I'll discuss option 2 with Alon.

Does that sound fair?

**Comment by DocCyblade** [comment]
@JedMeister - Last update tonight I promise! 

So looks like we can use the two existing [plugins](https://www.adminer.org/en/plugins/) as follows:

1. [login-servers](https://raw.github.com/vrana/adminer/master/plugins/login-servers.php) - Hardcode Server name (localhost) and Driver
2. [version-noverify](https://raw.github.com/vrana/adminer/master/plugins/version-noverify.php) - To disable the version checker

Doing this we would still need to implement by creating our own plugin.php

As I said above, we could if we want create our own version of the login form and TurnKey Brand it. We would just need to copy the login sever plug in and put the TurnKey plug in. 

Below are some examples I have running using the above. 

1. MySQL - https://krajr.tklapp.com:12322/tkl-mysql.php
2. PostgreSQL - https://krajr.tklapp.com:12322/tkl-psql.php
3. TKL Branded - https://krajr.tklapp.com:12322/tkl-branded.php

**Comment by DocCyblade** [comment]
Post some example scripts


Base File
---
```php
<?php
function adminer_object() {
        // required to run any plugin
        include_once "../plugins/plugin.php";

        // autoloader
        foreach (glob("../plugins/*.php") as $filename) {
                include_once $filename;
        }

        $plugins = array(
                // specify enabled plugins here
                
        );

        /* It is possible to combine customization and plugins:
        class AdminerCustomization extends AdminerPlugin {
        }
        return new AdminerCustomization($plugins);
        */

        return new AdminerPlugin($plugins);
}

// include original Adminer or Adminer Editor (usually named adminer.php)
include "./index.php";
```

MySQL
--- 
```php
...
        $plugins = array(
                // specify enabled plugins here
                new AdminerVersionNoverify,
                new AdminerLoginServers(array("localhost" => "MySQL on localhost"), "server" )
        );
...
```

PostgreSQL
--- 
```php
...
        $plugins = array(
                // specify enabled plugins here
                new AdminerVersionNoverify,
                new AdminerLoginServers(array("localhost" => "PostgreSQL on localhost"), "pgsql" )
        );
...
```


TurnKeyed Login Servers File
---
Usage would be the same as the above implementations above except we would have `use TKLAdminerLoginServers`

```php
<?php

class TKLAdminerLoginServers {
        var $servers, $driver;

        function __construct($servers, $driver = "server") {
                $this->servers = $servers;
                $this->driver = $driver;
        }

        function login($login, $password) {
                // check if server is allowed
                foreach ($this->servers as $key => $val) {
                        $servers = $val;
                        if (!is_array($val)) {
                                $servers = array($key => $val);
                        }
                        foreach ($servers as $k => $v) {
                                if ((is_string($k) ? $k : $v) == SERVER) {
                                        return;
                                }
                        }
                }
                return false;
        }

        function loginForm() {
                ?>
<table cellspacing="0">
<tr><th><?php echo lang('Server'); ?><td><input type="hidden" name="auth[driver]" value="<?php echo $this->driver; ?>"><select name="auth[server]"><?php echo optionlist($this->servers, SERVER); ?></select>
<tr><th><?php echo lang('Username'); ?><td><input id="username" name="auth[username]" value="<?php echo h($_GET["username"]);  ?>">
<tr><th><?php echo lang('Password'); ?><td><input type="password" name="auth[password]">
</table>
<p><input type="submit" value="<?php echo lang('Login'); ?>">
<?php echo checkbox("auth[permanent]", 1, $_COOKIE["adminer_permanent"], lang('Permanent login')) . "\n"; ?>
<br><br><br><br>TurnKey Linux - Adminer

<?php return true;
        }
```

**Comment by DocCyblade** [comment]
Just need to know what tag line you want at the bottom of the login screen. I can have the PR by this weekend. 

Speaking of PR, since most of this is all one change should it be one PR? it will affect multi files in common

**Comment by JedMeister** [comment]
I know I keep saying this, but only because it's true and you keep blowing me away by how awesome you are! :smile: You rock! 

As for tagline. Currently we tag the bottom of most appliances with "XYZ appliance - powered by TurnKey Linux" ("XYZ appliance" links to appliance page; "TurnKey Linux" links to website front page). But I guess that's hardly relevant because it's not an Adminer appliance! So maybe we should just go with "Adminer - powered by TurnKey Linux"? We could still link the "TurnKey Linux" to the home page, not sure what we should do with the Adminer. We could just leave it as text, link to Adminer upstream? Or we could create a doc page for Adminer on the website (/docs/adminer) and put a bit of info there?

TBH I'm happy for you to do what you think is best. If I were you, I'd be inclined to keep it simple.

Re PR:
A single PR is fine. Ideally each PR should only be achieving a single aim. However, sometimes that "single aim" might be "update for 14.2"! :smile: 

Commits on the other hand should generally be quite atomic. It's legitimate to update multiple files within a single commit; however they should ideally only be addressing one atomic point (e.g. fixing a single specific bug; applying a single specific feature request). 

OTOH editing a single file in multiple consecutive commits is also legitimate (and often preferred) when the point of each commit is different. 

To try to give some context; applying separate bugfixes in a single commit is bad; but doing a PR full of completely separate and individual bugfix commits for a specific appliance  is totally fair and reasonable.

I hope that makes sense. Keep in mind, it's a guide, not a law...

**Comment by DocCyblade** [comment]
Ok so I decided I would work on this a little more tonight... 

I was looking through common and saw that in the plans there are a lot of places lamp/lapp admin is in the plan. I think you said in order to use back ports we should ["pin it like this"](https://github.com/turnkeylinux-apps/asp-net-apache/blob/master/conf.d/main)  If that is the case:

1. Remove the theme overlay not needed with new version, we just need to symlink it in conf scripts
2. We would need to update all the conf scripts and use the pin there to update adminer and to link the new theme. 
3. Update web server configs to point to `tkl.index.php` a symlink to `tkl-mysql.index.php` or `tkl-pgsql.index.php` depending on the script in number 2

**Comment by JedMeister** [comment]
Yep that's a good pinning example. When testing and troubleshooting pinning, using `apt-cache policy <packagename>` is invaluable. It shows all available version of a package and what the relevant pinning values are. Keep in mind that values below 100 are irrelevant (an installed program will have a value of 100 so something with a value of 99 will never get updated). Values above 1000 are usable, but as a general rule should only ever be used when you explicitly wish to downgrade a package version. 500 is the common value given to packages which you want to be installed. Basically the package with the highest version number; and the highest pinning value (between 100 & 999) will be the installation candidate!

1. sounds good.
2 What about creating a new standalone generic `adminer` conf script? Then both the pinning, install and the theme linking could be handled there? > then we just add the new script to the common lamp/lapp make files. We may need to tweak some other individual appliance makefiles too, but that's ok. Ideally it'd be nice to just make both of those scripts depnd on a single `adminer` script but I'm not 100% sure about that. I'll need to ask Alon.
3. Perhaps we could integrate that step into the existing `adminer-mysql` and `adminer-postgres` common conf files?

**Comment by DocCyblade** [comment]
@JedMeister 

Always nice to chat with ya mate :-) Just a run down about what we talked about regarding adminer and how to best put the new version in to use. 

I am going to try and do one overlay, and one conf script that will detect pin the new version and update/install adminer then figure out what databases are installed and enable them in the drop down via the tcl-plugin file from the overlay. It will also detect what web server is running. If some how it multi-web servers are detected (not likely but you never know) it will detect lighty, then nginx, then apache in that order, where apache would win over the other. Once web server has been identified it will enable the correct etc file for said web server. 

I think I can make it pretty simple in that we use functions to detect the database servers and web servers, and the implementation as well to separate the logic for each. Should be easier to maintain and to test

If I can pull this off, it will allow easy adding of additional web servers and database servers to the script, and adding it to an app will be as simple as adding the overlay and the conf script. 

In the end I am going for complexity in one file vs a lot of simple files that would need to be referenced and that its self could be complex.

**Comment by DocCyblade** [comment]
I have started a new branch https://github.com/DocCyblade/tkl-common/tree/adminer-4.2-update to track my changes

**Comment by DocCyblade** [comment]
@JedMeister I have changed most of the files I think that need changed.  I need to do the pinning now. Thinking I would do that in the database specific scripts or create a adminer-common and move the theme and pinning there.

**Comment by JedMeister** [comment]
@DocCyblade - As you see fit. Although it's an extra file, perhaps having a "adminer-common" is a good idea?! I suspect that it would make what is going on more apparent to the unintiated. But bottom line mate, I know you well enough by now to know that whichever direction you go, you'll make sure it works. And obviously I'll do a double check over it as well.

**Comment by DocCyblade** [comment]
I am posting this here so other and my self need to pin something, have a good explanation documented. First off reading the manual is some times the best medicine, however you better be in a good frame of mind when reading `man apt_preferences` !

For this case I will be doing the following:

1. Create file at `/etc/apt/sources.list.d/backports.sources.list`
```
deb http://ftp.debian.org/debian jessie-backports main
```

2. Create file at `/etc/apt/preferences.d/adminer-backports.pref`

```
Explanation: Pin Adminer from backports
Package: adminer
Pin: release a=jessie-backports
Pin-Priority: 500
```


My Testing by running `apt-cache policy adminer`
----

#### Before Backports Repo Added:
```
adminer:
  Installed: (none)
  Candidate: 3.3.3-1
  Version table:
     3.3.3-1 0
        500 http://http.debian.net/debian/ jessie/main amd64 Packages
        100 /var/lib/dpkg/status
```

#### After Added Backports:
```
adminer:
  Installed: (none)
  Candidate: 3.3.3-1
  Version table:
     4.2.5-3~bpo8+1 0
        100 http://ftp.debian.org/debian/ jessie-backports/main amd64 Packages
     3.3.3-1 0
        500 http://http.debian.net/debian/ jessie/main amd64 Packages
        100 /var/lib/dpkg/status
```

#### After Adding Pinning
```
adminer:
  Installed: (none)
  Candidate: 4.2.5-3~bpo8+1
  Package pin: 4.2.5-3~bpo8+1
  Version table:
     4.2.5-3~bpo8+1 500
        100 http://ftp.debian.org/debian/ jessie-backports/main amd64 Packages
     3.3.3-1 500
        500 http://http.debian.net/debian/ jessie/main amd64 Packages
        100 /var/lib/dpkg/status
```

**Comment by JedMeister** [comment]
I have just been mucking around with backports myself and have a suggestion. You are currently explicitly using the main US Debian mirror. Obviously that isn't really an issue, however, it's probably superior to use the [redirection service](http://httpredir.debian.org/) as that will locate the best global mirror to use regardless of server location. I.e. use this line in the sources.list file:
````
deb http://httpredir.debian.org/debian jessie-backports main
````

I haven't done exhaustive testing, but it appears to work fine and I have no reason to believe that the backports repo would be any different to the redirector experience I've had with main, etc (which I've been personally using for ages).

**Comment by DocCyblade** [comment]
@JedMeister - sounds good! I'll make the change and start testing later tonight after We get back home

**Comment by DocCyblade** [comment]
@JedMeister - Went ahead and updated the code, FWIW I got that line from https://backports.debian.org/Instructions/ 

TBH I don't really know what exactly is going on in that line (I have an idea) but I think I'll read the man page on it :-)

**Comment by JedMeister** [comment]
My understanding is that the space separated entries after the URL are known as the "component" of the repo. I guess they are sort of like the individual "sub repos". 

It is totally legitimate to have them all on the one line (e.g. `jessie-backports main`) but IMO it is clearer to have each entry on separate line. It also allows for specific "components" to be commented out as per the default [TurnKey sources.list](https://github.com/turnkeylinux/common/blob/master/conf/turnkey.d/apt#L20-L22). `non-free` is commented out by default but can easily be uncommented with sed.

I suspect that having `main` declared twice is unlikely to cause anyone issues. However it could cause confusion if for some reason someone was trying to disable `main` and didn't realise it was also declared in the backports sources list.

**Comment by DocCyblade** [comment]
@JedMeister - Well without main being there it failed to build.

**Comment by JedMeister** [comment]
Hmm ok, so maybe it does need to be there?! Sry for that. Perhaps I'm missing something...

**Comment by DocCyblade** [comment]
@JedMeister 

Just batch built the following from build code using commit [df0f1d22c84b8976c4653eaf56da4d0b9ebc6b4f](https://github.com/DocCyblade/tkl-common/tree/df0f1d22c84b8976c4653eaf56da4d0b9ebc6b4f)

(Checked = Installed and Tested)
- [x] MySQL
- [x] Postgresql
- [x] LAMP
- [x] LAPP
- [x] Lighttpd-php-fastcgi
- [ ] Wordpress
- [ ] Odoo
- [ ] CakePHP
- [ ] Symphony
- [ ] Larvel
- [ ] Django



From the build log, you can see the pinning is working and from the screen shots you can see the theme, branding, and server lock in plugins are working.


#### MySQL Conf Script
```
fab-chroot build/root.patched --script /turnkey/fab/common/conf/adminer-mysql
Hit http://security.debian.org jessie/updates InRelease
Ign http://archive.turnkeylinux.org jessie-security InRelease          
Ign http://archive.turnkeylinux.org jessie InRelease                 
Hit http://security.debian.org jessie/updates/main amd64 Packages    
Hit http://archive.turnkeylinux.org jessie-security Release.gpg
Hit http://archive.turnkeylinux.org jessie Release.gpg                              
Hit http://security.debian.org jessie/updates/contrib amd64 Packages                
Hit http://archive.turnkeylinux.org jessie-security Release                             
Hit http://archive.turnkeylinux.org jessie Release                                                            
Hit http://security.debian.org jessie/updates/contrib Translation-en                                          
Hit http://archive.turnkeylinux.org jessie-security/main amd64 Packages                                                         
Hit http://security.debian.org jessie/updates/main Translation-en                                          
Hit http://archive.turnkeylinux.org jessie/main amd64 Packages                              
Get:1 http://httpredir.debian.org jessie-backports InRelease [166 kB]                          
Ign http://archive.turnkeylinux.org jessie-security/main Translation-en       
Ign http://archive.turnkeylinux.org jessie/main Translation-en
Ign http://httpredir.debian.org jessie InRelease
Get:2 http://httpredir.debian.org jessie-backports/main Translation-en [553 kB]
Hit http://httpredir.debian.org jessie Release.gpg            
Get:3 http://httpredir.debian.org jessie-backports/main amd64 Packages [806 kB]
Hit http://httpredir.debian.org jessie Release                                   
Hit http://httpredir.debian.org jessie/main amd64 Packages
Hit http://httpredir.debian.org jessie/contrib amd64 Packages
Hit http://httpredir.debian.org jessie/contrib Translation-en    
Hit http://httpredir.debian.org jessie/main Translation-en
Fetched 1525 kB in 2s (644 kB/s)
Reading package lists... Done
Before Adminer Update
adminer:
  Installed: 3.3.3-1
  Candidate: 4.2.5-3~bpo8+1
  Package pin: 4.2.5-3~bpo8+1
  Version table:
     4.2.5-3~bpo8+1 500
        100 http://httpredir.debian.org/debian/ jessie-backports/main amd64 Packages
 *** 3.3.3-1 500
        500 http://httpredir.debian.org/debian/ jessie/main amd64 Packages
        100 /var/lib/dpkg/status
Reading package lists... Done
Building dependency tree       
Reading state information... Done
Calculating upgrade... Done
The following packages will be upgraded:
  adminer
1 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.
Need to get 387 kB of archives.
After this operation, 621 kB of additional disk space will be used.
Get:1 http://httpredir.debian.org/debian/ jessie-backports/main adminer all 4.2.5-3~bpo8+1 [387 kB]
Fetched 387 kB in 0s (14.3 MB/s)
debconf: delaying package configuration, since apt-utils is not installed
(Reading database ... 28964 files and directories currently installed.)
Preparing to unpack .../adminer_4.2.5-3~bpo8+1_all.deb ...
Unpacking adminer (4.2.5-3~bpo8+1) over (3.3.3-1) ...
Setting up adminer (4.2.5-3~bpo8+1) ...
AFTER Adminer Update
adminer:
  Installed: 4.2.5-3~bpo8+1
  Candidate: 4.2.5-3~bpo8+1
  Package pin: 4.2.5-3~bpo8+1
  Version table:
 *** 4.2.5-3~bpo8+1 500
        100 http://httpredir.debian.org/debian/ jessie-backports/main amd64 Packages
        100 /var/lib/dpkg/status
     3.3.3-1 500
        500 http://httpredir.debian.org/debian/ jessie/main amd64 Packages

```

#### PostgreSQL Conf Script
```
fab-chroot build/root.patched --script /turnkey/fab/common/conf/adminer-pgsql
Hit http://security.debian.org jessie/updates InRelease
Ign http://archive.turnkeylinux.org jessie-security InRelease
Ign http://archive.turnkeylinux.org jessie InRelease                    
Hit http://archive.turnkeylinux.org jessie-security Release.gpg         
Hit http://archive.turnkeylinux.org jessie Release.gpg
Hit http://archive.turnkeylinux.org jessie-security Release
Hit http://archive.turnkeylinux.org jessie Release                                         
Hit http://archive.turnkeylinux.org jessie-security/main amd64 Packages                    
Hit http://security.debian.org jessie/updates/main amd64 Packages    
Hit http://archive.turnkeylinux.org jessie/main amd64 Packages                      
Hit http://security.debian.org jessie/updates/contrib amd64 Packages                    
Hit http://security.debian.org jessie/updates/contrib Translation-en                                          
Hit http://security.debian.org jessie/updates/main Translation-en                                                
Get:1 http://httpredir.debian.org jessie-backports InRelease [166 kB]                       
Ign http://archive.turnkeylinux.org jessie-security/main Translation-en                             
Ign http://archive.turnkeylinux.org jessie/main Translation-en                                      
Ign http://httpredir.debian.org jessie InRelease          
Hit http://httpredir.debian.org jessie Release.gpg  
Get:2 http://httpredir.debian.org jessie-backports/main Translation-en [553 kB]
Hit http://httpredir.debian.org jessie Release                               
Get:3 http://httpredir.debian.org jessie-backports/main amd64 Packages [806 kB]
Hit http://httpredir.debian.org jessie/main amd64 Packages                           
Hit http://httpredir.debian.org jessie/contrib amd64 Packages       
Hit http://httpredir.debian.org jessie/contrib Translation-en      
Hit http://httpredir.debian.org jessie/main Translation-en
Fetched 1525 kB in 2s (704 kB/s)
Reading package lists... Done
Before Adminer Update
adminer:
  Installed: 3.3.3-1
  Candidate: 4.2.5-3~bpo8+1
  Package pin: 4.2.5-3~bpo8+1
  Version table:
     4.2.5-3~bpo8+1 500
        100 http://httpredir.debian.org/debian/ jessie-backports/main amd64 Packages
 *** 3.3.3-1 500
        500 http://httpredir.debian.org/debian/ jessie/main amd64 Packages
        100 /var/lib/dpkg/status
Reading package lists... Done
Building dependency tree       
Reading state information... Done
Calculating upgrade... Done
The following packages will be upgraded:
  adminer
1 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.
Need to get 387 kB of archives.
After this operation, 621 kB of additional disk space will be used.
Get:1 http://httpredir.debian.org/debian/ jessie-backports/main adminer all 4.2.5-3~bpo8+1 [387 kB]
Fetched 387 kB in 0s (16.9 MB/s)
debconf: delaying package configuration, since apt-utils is not installed
(Reading database ... 29392 files and directories currently installed.)
Preparing to unpack .../adminer_4.2.5-3~bpo8+1_all.deb ...
Unpacking adminer (4.2.5-3~bpo8+1) over (3.3.3-1) ...
Setting up adminer (4.2.5-3~bpo8+1) ...
AFTER Adminer Update
adminer:
  Installed: 4.2.5-3~bpo8+1
  Candidate: 4.2.5-3~bpo8+1
  Package pin: 4.2.5-3~bpo8+1
  Version table:
 *** 4.2.5-3~bpo8+1 500
        100 http://httpredir.debian.org/debian/ jessie-backports/main amd64 Packages
        100 /var/lib/dpkg/status
     3.3.3-1 500
        500 http://httpredir.debian.org/debian/ jessie/main amd64 Packages
```

#### Screen shot of Adminer Login Screen on built LAMP
<img width="941" alt="lamp" src="https://cloud.githubusercontent.com/assets/9731860/22809137/26fb04a4-eefe-11e6-9125-61f57e7b7036.png">



#### Screen shot of Adminer Login Screen on built LAPP
<img width="987" alt="lapp" src="https://cloud.githubusercontent.com/assets/9731860/22809142/2f9745b4-eefe-11e6-921d-5d5fe441f4ad.png">



#### Screen shot of Adminer Login Screen on built Lighttpd-php-fastcgi
<img width="1092" alt="lighty" src="https://cloud.githubusercontent.com/assets/9731860/22809146/34af5186-eefe-11e6-89b9-0597836e362f.png">

**Comment by l-arnold** [comment]
Looks like a piece of Art to me Ken!  
I assume it all worked out too.

Adminer has been working well for me as a point of note.

**Comment by JedMeister** [comment]
Awesome work @DocCyblade! 

And I agree @l-arnold! The work Ken has done on the integration is superb!

BTW I'm glad to hear that you are settling in to Adminer. I recall a few grumblings when we first switched... :smile: FWIW this newer version apparently has much better Postgres support and is better overall generally. I'll look forward to your feedback when we get our next release out the door...

**Comment by JedMeister** [comment]
While testing and bugfixing the LE confconsole plugin (over [here](https://github.com/turnkeylinux/tracker/issues/766#issuecomment-284548224)) I thought I'd kill 2 birds with one stone and built an Nginx server with the new Adminer included too. 

It all seems to work fine (as I expected) however I did notice that on Nginx, it doesn't look quite the same as your screenshots @DocCyblade?! It seems to be missing the "branding" and has the extra (unneeded boxes) and doesn't include the hostname as yours appears to?!

FWIW I am using a new branch I've created called [14.2-testing](https://github.com/JedMeister/common/tree/14.2-testing) which (should) include all the changes planned for v14.2. I have merged your [adminer-4.2-update](https://github.com/DocCyblade/tkl-common/tree/adminer-4.2-update) branch into it. Is there something that I'm missing? Or have I discovered something you overlooked/missed? TBH I find that hard to believe! :smile:
